--- a/OptionParser.h
+++ b/OptionParser.h
@@ -222,7 +222,7 @@
 class OptionGroup : public OptionParser {
   public:
     OptionGroup(const OptionParser& p, const std::string& t, const std::string& d = "") :
-      _parser(p), _title(t), _group_description(d) {}
+      _title(t), _group_description(d) {}
     virtual ~OptionGroup() {}
 
     OptionGroup& title(const std::string& t) { _title = t; return *this; }
@@ -231,7 +231,6 @@
     const std::string& group_description() const { return _group_description; }
 
   private:
-    const OptionParser& _parser;
     std::string _title;
     std::string _group_description;
 };
